src/test/java/org/alfresco/repo/site/SiteServiceImplMoreTest.java

```diff
@@ -1214,4 +1214,81 @@ public TestData(Map<String, String> userNameToRoleMap, NodeRef testDocNodeRef)
+
+    /**
+     * REPO-2811 / ALF-21924: "Internal Server Error" when listSites is invoked
+     * for users belonging to a deleted Site.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testListMembershipOnSitesDifferentCase() throws Exception
+    {
+        final String userSiteOwner = "UserSiteOwner";
+        final String userSiteCollaborator = "UserSiteCollaborator";
+        final String userPrefix = "test";
+
+        // Create the users
+        TRANSACTION_HELPER.doInTransaction(new RetryingTransactionCallback<Void>()
+        {
+            public Void execute() throws Throwable
+            {
+                createUser(userSiteOwner, userPrefix);
+                createUser(userSiteCollaborator, userPrefix);
+                return null;
+            }
+        });
+
+        // Create site
+        final String id = Long.toString(System.currentTimeMillis());
+        final String siteShortName = "testsite-" + id;
+
+        log.debug("Creating test site called: " + siteShortName);
+
+        perMethodTestSites.createSite("sitePreset", siteShortName, null, null, SiteVisibility.PUBLIC, userSiteOwner);
+
+        // Add Site COLLABORATOR
+        TRANSACTION_HELPER.doInTransaction(new RetryingTransactionCallback<Void>()
+        {
+            public Void execute() throws Throwable
+            {
+                // make userSiteCollaborator a member of the site
+                SITE_SERVICE.setMembership(siteShortName, userSiteCollaborator, SiteModel.SITE_COLLABORATOR);
+
+                return null;
+            }
+        });
+
+        // Delete site
+        TRANSACTION_HELPER.doInTransaction(new RetryingTransactionCallback<Map<String, String>>()
+        {
+            public Map<String, String> execute() throws Throwable
+            {
+                log.debug("About to delete site.");
+                SITE_SERVICE.deleteSite(siteShortName);
+                log.debug("Site deleted.");
+
+                return null;
+            }
+        });
+
+        /*
+         * Create site with different case than the deleted one.
+         * This is possible since the site id validation is case insensitive only for existing sites
+         * while sites in trashcan are validated using authorities which are case sensitive
+         */
+        final String siteShortName1 = "TesTsite-" + id;
+        log.debug("Creating test site called: " + siteShortName1);
+
+        perMethodTestSites.createSite("sitePreset", siteShortName1, null, null, SiteVisibility.PUBLIC, userSiteOwner);
+
+        // Check that user membership can be retrieved and user is not a member of existing site
+        List<SiteMembership> members = SITE_SERVICE.listSiteMemberships(userSiteCollaborator, 0);
+        assertNotNull(members);
+        assertEquals(0, members.size());
+
+        members = SITE_SERVICE.listSiteMemberships(userSiteCollaborator, 50);
+        assertNotNull(members);
+        assertEquals(0, members.size());
+    }

```

Comment:
Prior to testing if listing the site memberships works, add the user to a _good_ site, to also test that the matching of the sites retrieved from authorities, still works.